### ISSUE TITLE
Support TypeScript async import

### ISSUE THREAD

**Issue by jamonholmgren** [infinitered/gluegun#342]
Per [this comment on #337](https://github.com/infinitered/gluegun/issues/337#issuecomment-362420041), we'd like to use TypeScript's built-in dynamic `import` rather than commonjs `require`.

This is the relevant bit of code:

https://github.com/infinitered/gluegun/blob/69dd58fc64a61d44e1393dd7fb059ced7f64b5ad/src/loaders/module-loader.ts#L10-L14

## Should we switch to dynamic import?

TypeScript's [dynamic import](https://blog.mariusschulz.com/2018/01/14/typescript-2-4-dynamic-import-expressions) expressions would work here. _However_, this would require somewhat of an internal refactor, because while `require` is synchronous, dynamic `import` is asynchronous.

This means that we would need to defer loading any plugins, extensions, and commands until we `create()` the runtime -- and this would need to change to an async function, a breaking change.

```js
// current
const cli = build()
  .brand('mycli')
  .src(__dirname)
  .create()

const context = await cli.run()

// with dynamic import
const cli = await build()
  .brand('mycli')
  .src(__dirname)
  .create()

const context = await cli.run()
```

As you can see, there's not a lot of difference there, but it _is_ a breaking change.

Internally, we'd need to change several functions/methods to async, including `Builder#create`, `loadModule`, `loadCommandFromFile`, `loadExtensionFromFile`, `loadPluginFromDirectory`. We'd also have to defer loading any plugins until you call `create` (which, incidentally, goes the opposite way from [my previous "simplification" refactor](https://github.com/infinitered/gluegun/pull/307/files#diff-7664f037f64125ca234fb82ed2d46cb4L26). 🤦‍♂️ 

It's somewhat mapped out in my head. I'm not really looking forward to the refactor, though, to be honest. And since it works okay right now, I'm not super motivated.

**However, I am a bit worried about breaking the API later, when Gluegun has more traction. So I propose we change `Builder#create()` to an async function right now, even though it doesn't need it currently, to prepare for a future point when we move to dynamic imports.**

Thoughts?

**Comment by skellock** [comment]
### TL;DR

It's beautiful just the way it is.

### Good for the Web is Good for the CLI?

The splitting scenario is wonderful for the web, I don't doubt that.

For us, it *seems* like a good fit too.  For example:

* we bootstrap the `gluegun` runtime, but then only run 1 command.
* `solidarity` has multiple plugins (`react-native` and `elixir`) but might only need 1 set.

I get the allure of wanting to do this.

### Performance is the Selling Point?

It's already fast.

Yes, it could be faster. We're venturing really deep in diminished returns territory though.

IMH-(strawman)-O, we could cut down the overhead by an order of magnitude by switching to `rust` or `go`.  I hope we don't do that either.

### Stable APIs

I wouldn't switch `create()` to be `async`.

We don't need it.

If we need `async` in `gluegun@3`, then let's switch it up at that point.  There's lots of techniques to do this safely (alternative name, alternative import, alternative lib (lulz) etc).

In the mean time you can do your own `async` by deferring your `.create()`. This buys you time to put your own dynamic plugin loading in.

### Caveat

The 🤓 in me is really disappointed.  Dynamic imports look really neat and I'd love to try it.

**Comment by jamonholmgren** [comment]
![excellent point](https://media1.tenor.com/images/e2b6c354cedaabd81e0adb0d4fdefc6c/tenor.gif?itemid=8337027)
